// @redocly/openapi-cli/dist/error/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getReferencedFrom", {
  enumerable: true,
  get: function () {
    return _default2.getReferencedFrom;
  }
});
exports.createYAMLParseError = exports.default = exports.messageHelpers = void 0;

var _path = require("path");

var _default2 = _interopRequireWildcard(require("./default"));

var _yaml = require("../yaml");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const mutuallyExclusiveFieldsMessageHelper = fieldNames => `Fields ${fieldNames.map(el => `'${el}'`).join(', ')} are mutually exclusive.`;

const fieldTypeMismatchMessageHelper = desiredType => `This field must be of ${desiredType} type.`;

const missingRequiredField = fieldName => `The field '${fieldName}' must be present on this level.`;

const fieldNotAllowedMessageHelper = (fieldName, definitionName) => `The field '${fieldName}' is not allowed in ${definitionName}. Use "x-" prefix or custom types to override this behavior.`;

const messageHelpers = {
  mutuallyExclusiveFieldsMessageHelper,
  fieldTypeMismatchMessageHelper,
  missingRequiredField,
  fieldNotAllowedMessageHelper
};
exports.messageHelpers = messageHelpers;
var _default = _default2.default;
exports.default = _default;

const createYAMLParseError = (e, ctx, resolvedPath, source, root = false) => ({
  message: `${e.name}: ${e.reason}`,
  path: root ? [] : Array.from(ctx.path),
  referencedFrom: root ? null : (0, _default2.getReferencedFrom)(ctx),
  location: {
    startLine: e.mark.line + 1,
    startCol: e.mark.column
  },
  codeFrame: (0, _yaml.getCodeFrameForLocation)(e.mark.position - 1, e.mark.position, source, e.mark.line + 1),
  value: null,
  file: (0, _path.relative)(process.cwd(), resolvedPath),
  severity: 4,
  enableCodeframe: ctx.enableCodeframe !== undefined ? ctx.enableCodeframe : true
});

exports.createYAMLParseError = createYAMLParseError;
